Faster BitArray packing and related operations

carlobaldassi opened the pull request:
This patch accelerates the following operations:
- broadcasting operations which return BitArrays (e.g. `.==`)
- packing operations of Arrays into BitArrays (with a specialized case for `Array{Bool}`):
  - extends the signature of `copy!` to `copy!(dest::BitArray, doffs::Integer, src::Array, soffs::Integer, n::Integer)` and to `copy!(dest::BitArray, src::Array)`
  - uses the new `copy!` in `convert` operations
  - factors out an `unsafe_copy!` version like for Arrays
- indexing of BitArrays with contiguous first indexing expression:
  - adds `Colon` in the signatures such that `b[:]` is as fast as `b[1:end]`
  - removes `Bool` from scalar signatures such that `b[:] = 1` is as fast as `b[:] = true`
  - also implements some TODO's which were left in the multidimensional.jl code

Also adds tests and a small comment to explain the logic of the faster packing trick.

For reference, the faster broadcasting is all due to avoiding a loop of the form `0:8:63` which required a relatively expensive modulus operation. To my surprise, that represented a large portion of the time spent when packing.

Some code may be generalizable to AbstractArrays but the `copy!(src, soffs, dest, doffs, n)` signature inherently uses linear indexing, so it's not totally straightforward.

carlobaldassi committed:
Fix Hacker's Delight URL in comment + cosmetics

carlobaldassi committed:
bitarray tests: add consistency check

also slightly increase 4D test case size

carlobaldassi committed:
bitarray multidim. scalar setindex!: remove ::Bool

such that specialized fast indexing functions catch
assignments with other types

carlobaldassi committed:
bitarray setindex!: implement missing shape check

also fix some other todos

carlobaldassi committed:
bitarray indexing: add Colon to signatures

so that specialized indexing catches also b[:], not only b[1:end]

carlobaldassi committed:
Faster Array->BitArray packing

* move the function which was in broadcast.jl, generalize it and
  rename it
* also add a version for Array, not only Array{Bool}
* rename  the auxiliary function bpack->pack8bools, add comment
* use the fast packing functions in copy! and convert (therefore,
  also in bitpack)
* split out an unsafe_copy! version
* add tests

carlobaldassi committed:
bitarray multid. setindex: add Array in signatures

uses faster packing scheme

carlobaldassi commented:
`runbenchmarks(ALL, vs = "JuliaLang/julia:master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/15863#issuecomment-209557359) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e46754137c775df7a7594166e39120d36ae0c7bc/5c410f0/5c410f0_vs_1b4618e.md). cc @jrevels

timholy commented:
> For reference, the faster broadcasting is all due to avoiding a loop of the form 0:8:63 which required a relatively expensive modulus operation. To my surprise, that represented a large portion of the time spent when packing.

I don't understand how, given https://github.com/JuliaLang/julia/blob/14230288717ec0ebd92df1956e1ec46ef5a3d464/base/range.jl#L391-L395. Can you explain further?

carlobaldassi commented:
> I don't understand how, given https://github.com/JuliaLang/julia/blob/14230288717ec0ebd92df1956e1ec46ef5a3d464/base/range.jl#L391-L395. Can you explain further?

Sorry, my description was sloppy. The modulus is called by the range constructor, here: https://github.com/JuliaLang/julia/blob/14230288717ec0ebd92df1956e1ec46ef5a3d464/base/range.jl#L63

In current master, it is called from here when broadcasting: https://github.com/JuliaLang/julia/blob/14230288717ec0ebd92df1956e1ec46ef5a3d464/base/broadcast.jl#L121

In this patch, it was changed to this: https://github.com/JuliaLang/julia/blob/560d76dd1b4e2560b1608976eecef14fabb3fdc6/base/bitarray.jl#L256-L257

Also, "large portion of the time" may have been a bit of an overstatement. It is definitely noticeable when profiling though.

BTW I need to look into the possible performance regressions reported by benchmarking, the "sparse arrays" one looks suspicious.

timholy commented:
Ah, thanks for explaining. I forgot about that. It is amazing how frequently the slowness `div`/`rem` ends up [ruining my life](http://julialang.org/blog/2016/03/arrays-iteration#subarrays-a-case-study) :smile:.

mbauman commented on base/multidimensional.jl:
```diff
@@ -579,7 +579,7 @@ end
         storeind = 1
         Xc, Bc = X.chunks, B.chunks
-        idxlens = index_lengths(B, I0, I...) # TODO: unsplat?
+        idxlens = @ncall $N index_lengths B I0 d->I[d]
```
I'm pretty sure the splat here is fine — `index_lengths` always inlines so it should be equivalent to the `@ncall` version.  But it's definitely easier to reason about the performance of `@ncall`, so this is fine.

mbauman commented on base/multidimensional.jl:
```diff
@@ -617,32 +617,33 @@ end
 end
 
-@inline function setindex!(B::BitArray, X::BitArray, I0::UnitRange{Int}, I::Union{Int,UnitRange{Int}}...)
+@inline function setindex!(B::BitArray, X::BitArray, I0::UnitRange{Int}, I0::Union{Colon,UnitRange{Int}}, I::Union{Int,UnitRange{Int},Colon}...)
```
Will the repeated `I0` here cause a bootstrap failure?  If so - this is a friendly reminder to squash. :)

carlobaldassi replied:
Fixed, thanks for spotting this.

carlobaldassi committed:
Optimize Array->BitArray a little further

specialized Array{Real} version

carlobaldassi commented:
The reported performance regressions were apparently flukes. I fixed an intermediate commit and added another optimization. This should be good to go after tests pass.
